### ISSUE TITLE
Gradient clipping runs on the accumulated mean, so one micro-step in 128 can steer the update

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#201]
Gradient clipping is applied to the **mean of 128 accumulated micro-steps**, never to a
micro-step. The per-micro-step gradient distribution is heavy-tailed enough that this
makes the clip almost useless: it bounds the size of the update while letting a single
outlier choose its direction.

## Measured

9,000 micro-steps replayed from the step-11,583 checkpoint of the #157 run, recording
every per-micro-step `grad_norm` individually (the trainer only ever logs their mean):

```
p50           11.4
p75           14.8
p90           24.7
p95           62.8
p99          603.5
p99.5      1,349.1
p99.9      7,617.3
max       75,330.7        <- 6,610x the median
mean          56.3        <- the only number the run records

norms > 100:   3.61%
norms > 1000:  0.69%
norms > 10000: 0.04%
```

**The top 1% of micro-steps carry 64.5% of the total gradient mass; the top 0.1% carry
30.1%.** 99% of training is healthy at a norm of ~11-60.

## Why the current clip cannot help

`trm/train/optimizers.py:23`:

```python
optax.MultiSteps(
    optax.chain(
        optax.clip_by_global_norm(1.0),
        optax.adamw(...)
```

`MultiSteps` wraps the chain, so `clip_by_global_norm` runs once per *optimizer* step, on
the accumulated mean. One micro-step at 75,330 among 127 at ~11 produces a mean whose
direction is ~99% the outlier's. The clip then rescales that to norm 1.0 — a
correctly-sized step pointing almost entirely where one pathological micro-step wanted to
go. Every telemetry reading looks healthy while the model is dragged off.

This is what destroyed the run at opt step 11,650 (val CE 10.73 against
ln(50304) = 10.83 — a uniform distribution).

## The accidental protection we removed

The same 9,000-micro-step replay **did not collapse**, and passed cleanly through the
step where the live run died. The difference: 27 of its micro-steps overflowed f16 and
were skipped — and the ones that overflow are exactly the monsters.

Dynamic loss scaling (#199) had been acting as an unintentional outlier filter. In the
live run its scale had been driven down to 1, so nothing overflowed, nothing was skipped,
and every outlier landed at full strength. That explains both why the run was unusually
healthy for 500 steps after #199 landed (mean norms of 20-29, the best in days) and why
it then died.

Strongly supported rather than proven — but it is consistent with every observation.

## Fix

Clip each micro-step before it is accumulated, doing deliberately and reliably what the
overflow was doing by accident.

The threshold should be **adaptive, not a constant**: gradient magnitudes shrink as a run
converges, so a number calibrated at step 11,000 is miscalibrated by step 25,000. Track a
running scale of the typical micro-step norm and clip at a multiple of it, updating that
scale from clipped values so outliers cannot raise their own bar.

At 8x the observed median this touches ~3.6% of micro-steps and leaves 96.4% bit-identical
— which is what makes it safe to introduce mid-run.

The existing `clip_by_global_norm(1.0)` on the accumulated mean stays; the two guard
different things.

**Not novel** — per-microbatch clipping is standard practice. The PR is the record
(doctrine rule 5). What is worth keeping is the measurement above: this project had no
per-micro-step gradient telemetry at all, which is why the failure was invisible for
2,000 steps.

## Related

- #199 — loss scaling; its scale collapsing to 1 is what exposed this.
- #82 — the zero-grad instrument. Same shape of lesson: the mean hid the tail.
